store/tikv/region_request.go

```diff
@@ -213,8 +213,16 @@ func (s *RegionRequestSender) onRegionError(bo *Backoffer, ctx *RPCContext, regi
 
 	if staleEpoch := regionErr.GetStaleEpoch(); staleEpoch != nil {
 		log.Debugf("tikv reports `StaleEpoch`, ctx: %v, retry later", ctx)
-		err = s.regionCache.OnRegionStale(ctx, staleEpoch.NewRegions)
-		return false, errors.Trace(err)
+
+		if err = s.regionCache.OnRegionStale(ctx, staleEpoch.NewRegions); err != nil {
+			return false, errors.Trace(err)
+		}
+		// retry if error is StaleEpoch, and it will recover soon
+		if err = bo.Backoff(boRegionStale, errors.Errorf("stale epoch: %v, ctx: %v", staleEpoch, ctx)); err != nil {
```

Comment:
You should not retry inside the `RegionRequestSender`, because the request may not belong to the region anymore. You should find the region for the request and retry outside.